### PR TITLE
Reflection_Engine: Consolidate run extension method

### LINKED ISSUE #1865: Reflection_Engine: Consolidate RunExtensionMethod
#### Description:

A few things that should be improved for the RunExtensionMethod:
- Make it thread safe (Partially taken care of by https://github.com/BHoM/BHoM_Engine/pull/1857 and to be fully handled in https://github.com/BHoM/BHoM_Engine/issues/1863)
- Make run extension method able to handle generic methods, similar to exercises done in the UI in https://github.com/BHoM/BHoM_UI/pull/222)
- Look into if delegates can be stored instead of method info, to increase performance, similar to what is being done in https://github.com/BHoM/BHoM_Engine/blob/master/Reflection_Engine/Query/DistinctProperties.cs

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #1865 

Multiple fixes and improvements to the Run extension method. See changelog

### Test files

Testing scripts calling methods making use of RunExtensionMethod still work:

https://burohappold.sharepoint.com/:f:/s/BHoM/EvPhiod3VEVOluLAwReZmD4BPVEzu1U-DibAAAg10tBOjQ?e=mi8zDD

I have added a few test methods locally and tested that they do work appropriately, but do not want to push them to this PR. Example method in Analytical_Engine should make the BEnv panels visable (or any IPanel for that matter):

```
namespace BH.Engine.Analytical
{
    public static partial class Query
    {
        /***************************************************/
        /**** Public Methods                            ****/
        /***************************************************/

        [Description("")]
        [Input("", "")]
        [Output("", "")]
        public static PlanarSurface Geometry<TEdge, TOpening>(this IPanel<TEdge, TOpening> panel) where TEdge : IEdge where TOpening : IOpening<TEdge>
        {
            return new PlanarSurface(
                    Engine.Geometry.Compute.IJoin(panel.ExternalEdges.Select(x => x.Curve).ToList()).FirstOrDefault(),
                    panel.Openings.SelectMany(x => Engine.Geometry.Compute.IJoin(x.Edges.Select(y => y.Curve).ToList())).Cast<ICurve>().ToList());
        }

        /***************************************************/
    }
}
```

### Changelog

- Improve support of generics for RunExtensionMethod
- Add locks for accessing stored methods,  making the method thread safe
- Storing compiled Functions instead of MethodInfo to improve performance

### CODE REVIEW

File: Reflection_Engine/Compute/RunExtensionMethod.cs [hunk 1]
```diff
@@ -36,72 +38,131 @@ public static partial class Compute
...
+        {
+            lock (m_RunExtensionMethodLock)
+            {
+                return m_PreviousInvokedMethods.ContainsKey(key);
+            }
+        }
+
+        /***************************************************/
+
+        [Description("Gets a previously extracted method from the stored methods. Put in its own method to simplify the use of locks to provide thread safety.")]
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Reflection_Engine/Compute/RunExtensionMethod.cs:139]: Reason I wanted to keep both was that I am storing nulls for cases where no method is found. This is to avoid searching again and again when no method exist for a particular set of inputs. Say for example displaying geometry, it is being called for all BHoMObejct types. Without storing null, it will keep on searching again and again for an applicable method, all in vain.

This means "null" is a valid answer for "ContainsKey", hence can't use that to determaine if the method has previously been attempted to be extracted or not.

TL;DR

`MethodPreviouslyExtracted` checks if an attempt has been made to extract the method.
`GetStoredExtensionMethod` returns is if existing. Can be null for a case when an attempt has been made before, but failed to find anything -> No point trying again. (nulls are filtered out by the maybe not so visible `?` sign on the line it is being called)

File: Reflection_Engine/Compute/RunExtensionMethod.cs [hunk 2]
```diff
@@ -36,72 +38,131 @@ public static partial class Compute
...
+            if (parameters == null || parameters.Length == 0 || parameters.Any(x => x == null) || string.IsNullOrWhiteSpace(methodName))
+                return null;
 
-            // If the method has been called before, just use that
-            string name = methodName + parameters.Select(x => x.GetType().ToString()).Aggregate((a,b) => a+b);
+            //Get type of first argument, to be used for first method extraction filtering
+            Type type = parameters[0].GetType();
+
+            //Construct key used to store/extract method
+            string name = methodName + parameters.Select(x => x.GetType().ToString()).Aggregate((a, b) => a + b);
```
[the remark this answers is not in the document]
Comment (adecler) [reply Reflection_Engine/Compute/RunExtensionMethod.cs:81]: I personally disagree. It would also be good to still leave some personal expression freedom to each developer.

[the remark this answers is not in the document]
Comment (pawelbaran) [reply Reflection_Engine/Compute/RunExtensionMethod.cs:81]: Sure, that is why I used _could_ term 🙊 that was only my intuitive thought when looking at the code, my comment was not meant to request change on this particular one.

File: Reflection_Engine/Compute/RunExtensionMethod.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Reflection_Engine/Compute/RunExtensionMethod.cs:139]: Right, makes perfect sense.

File: Reflection_Engine/Compute/RunExtensionMethod.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Reflection_Engine/Compute/RunExtensionMethod.cs:81]: Do not really mind either, but keeping as is for now.

File: Reflection_Engine/Query/IsAssignableToGenericType.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            foreach (var inter in interfaceTypes)
+            {
+                //Check if the generic type definition for any of the interfaces matches the generic type
+                if (inter.IsGenericType && inter.GetGenericTypeDefinition() == genericType)
+                    return true;
+            }
+
+            //Check if given type contains a base type
+            Type baseType = givenType.BaseType;
+            if (baseType == null) return false;
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Reflection_Engine/Query/IsAssignableToGenericType.cs]: done

File: Reflection_Engine/Compute/RunExtensionMethod.cs [hunk 4]
```diff
@@ -36,72 +38,131 @@ public static partial class Compute
...
 
-                m_PreviousInvokedMethods[key] = method;
-                return method.Invoke(null, new object[] { target }.Concat(parameters).ToArray());
+                MethodInfo finalMethod = method;
+
+                //If method is generic, make sure the appropriate generic arguments are set
+                if (method.IsGenericMethod)
+                    finalMethod = method.MakeGenericFromInputs(parameters.Select(x => x.GetType()).ToList());
+
+                //TUrn the MethodInfo to a compiled function, store it and finally call it
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Reflection_Engine/Compute/RunExtensionMethod.cs]: Done

File: Reflection_Engine/Compute/RunExtensionMethod.cs [hunk 5]
```diff
@@ -36,72 +38,131 @@ public static partial class Compute
...
                 {
-                    if (!paramInfo[i + 1].ParameterType.IsAssignableFrom(parameters[i].GetType()))
+                    if(!paramInfo[i].ParameterType.IsAssignableFromIncludeGenerics(parameters[i].GetType()))
                     {
+                        //Parameter does not match, abort for this method
                         matchingTypes = false;
                         break;
                     }
                 }
                 if (!matchingTypes)
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Reflection_Engine/Compute/RunExtensionMethod.cs:105]: Done

File: Reflection_Engine/Compute/RunExtensionMethod.cs [hunk 6]
```diff
@@ -36,72 +38,131 @@ public static partial class Compute
...
                     continue;
 
-                m_PreviousInvokedMethods[key] = method;
-                return method.Invoke(null, new object[] { target }.Concat(parameters).ToArray());
+                MethodInfo finalMethod = method;
+
+                //If method is generic, make sure the appropriate generic arguments are set
+                if (method.IsGenericMethod)
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Reflection_Engine/Compute/RunExtensionMethod.cs]: good catch. Fixed and updated

**Comment by IsakNaslundBh** [comment]
Thanks both
